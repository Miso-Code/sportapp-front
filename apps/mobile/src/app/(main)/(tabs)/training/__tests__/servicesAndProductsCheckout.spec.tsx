import React from 'react'
import renderer, { ReactTestRenderer, act } from 'react-test-renderer'
import { router, useLocalSearchParams } from 'expo-router'
import {
	useAlertStore,
	useBusinessPartnerStore,
	useUserStore
} from '@sportapp/stores'

import ServicesAndProductsCheckout from '../servicesAndProductsCheckout'

jest.mock('expo-router')

jest.mock('react-native-paper', () => {
	const textInput = jest.requireActual('react-native').TextInput
	textInput.Icon = jest.requireActual('react-native-paper').TextInput.Icon
	return {
		...jest.requireActual('react-native-paper'),
		Portal: jest.fn(({ children }) => children),
		TextInput: textInput //prevents animation issues on jest
	}
})
jest.mock('@/components/KeyboardAvoidingDialog', () => {
	const native = jest.requireActual('react-native')
	return jest.fn(({ children, ...props }) => (
		<native.View {...props} testID='modalProcessing'>
			{children}
		</native.View>
	))
})
jest.mock('@/components/ProductServiceCard', () => {
	const native = jest.requireActual('react-native')
	return jest.fn(({ children, ...props }) => (
		<native.View {...props} testID='productCard'>
			{children}
		</native.View>
	))
})
jest.mock('@sportapp/stores', () => ({
	useAlertStore: jest.fn(),
	useBusinessPartnerStore: jest.fn(),
	useUserStore: jest.fn()
}))
describe('ServicesAndProductsCheckout', () => {
	let component: ReactTestRenderer

	beforeEach(() => {
		;(useLocalSearchParams as jest.Mock).mockReturnValue({
			quantity: 10,
			product: 1
		})
		;(useAlertStore as unknown as jest.Mock).mockReturnValue({
			setAlert: jest.fn()
		})
		;(useBusinessPartnerStore as unknown as jest.Mock).mockReturnValue({
			productToCheckout: {
				product_id: 'product_id',
				category: 'category',
				name: 'name',
				url: 'url',
				price: 100,
				payment_type: 'payment_type',
				payment_frequency: 'payment_frequency',
				image_url: 'image_url',
				description: 'description',
				active: true
			},
			purchaseProduct: jest.fn().mockResolvedValue({
				transaction_id: 'transaction_id',
				transaction_status: 'complete',
				transaction_date: 'transaction_date',
				message: 'message'
			})
		})
		;(useUserStore as unknown as jest.Mock).mockReturnValue({
			getProfile: jest.fn().mockResolvedValue({
				first_name: 'John',
				last_name: 'Doe',
				email: 'john.doe@example.com'
			})
		})
		component = renderer.create(<ServicesAndProductsCheckout />)
		jest.useFakeTimers()
	})

	afterEach(() => {
		component.unmount()
		jest.useRealTimers()
		jest.resetAllMocks()
	})

	it('should render correctly', () => {
		expect(component.toJSON()).toMatchSnapshot()
	})

	it('should redirect back if no product is selected', async () => {
		;(useBusinessPartnerStore as unknown as jest.Mock).mockReturnValue({
			productToCheckout: undefined
		})
		await act(async () => {
			component.update(<ServicesAndProductsCheckout />)
			await Promise.resolve()
		})
		expect(router.back).toHaveBeenCalled()
	})

	describe('Input Quantity', () => {
		it('should assign quantity params to input', () => {
			const quantityInput = component.root.findByProps({
				testID: 'quantity'
			})
			expect(quantityInput.props.value).toBe('10')
		})

		it('should update the quantity on input change', () => {
			const quantityInput = component.root.findByProps({
				testID: 'quantity'
			})
			act(() => {
				quantityInput.props.onChangeText('5')
			})
			expect(quantityInput.props.value).toBe('5')
		})
	})
	describe('Input Card Number', () => {
		it('should assign undefined card number params to input', () => {
			const cardNumberInput = component.root.findByProps({
				testID: 'cardNumber'
			})
			expect(cardNumberInput.props.value).toBe(undefined)
		})

		it('should update the card number and format on input change', () => {
			const cardNumberInput = component.root.findByProps({
				testID: 'cardNumber'
			})
			act(() => {
				cardNumberInput.props.onChangeText('1234567890123456')
			})
			expect(cardNumberInput.props.value).toBe('1234 5678 9012 3456')
		})

		it('should not allow more than 16 digits', async () => {
			const cardNumberInput = component.root.findByProps({
				testID: 'cardNumber'
			})
			act(() => {
				cardNumberInput.props.onChangeText('12345')
			})
			act(() => {
				cardNumberInput.props.onChangeText('1234567890123456789012345')
			})
			expect(cardNumberInput.props.value).toBe('1234 5678 9012 3456')
		})
	})
	describe('Input Card Holder', () => {
		it('should assign undefined card holder params to input', () => {
			const cardHolderInput = component.root.findByProps({
				testID: 'cardHolder'
			})
			expect(cardHolderInput.props.value).toBe('')
		})

		it('should update the card holder on input change to uppercase', () => {
			const cardHolderInput = component.root.findByProps({
				testID: 'cardHolder'
			})
			act(() => {
				cardHolderInput.props.onChangeText('John Doe')
			})
			expect(cardHolderInput.props.value).toBe('JOHN DOE')
		})
	})
	describe('Input Expiration Date', () => {
		it('should assign undefined expiration date params to input', () => {
			const expirationDateInput = component.root.findByProps({
				testID: 'expirationDate'
			})
			expect(expirationDateInput.props.value).toBe(undefined)
		})

		it('should update the expiration date on input change with format', () => {
			const expirationDateInput = component.root.findByProps({
				testID: 'expirationDate'
			})
			act(() => {
				expirationDateInput.props.onChangeText('1223')
			})
			expect(expirationDateInput.props.value).toBe('12/23')
		})

		it('should not allow more than 4 digits', async () => {
			const expirationDateInput = component.root.findByProps({
				testID: 'expirationDate'
			})
			act(() => {
				expirationDateInput.props.onChangeText('12345678')
			})
			expect(expirationDateInput.props.value).toBe('12/34')
		})
	})
	describe('Input CVV', () => {
		it('should assign undefined cvv params to input', () => {
			const cvvInput = component.root.findByProps({
				testID: 'cvv'
			})
			expect(cvvInput.props.value).toBe('')
		})

		it('should update the cvv on input change', () => {
			const cvvInput = component.root.findByProps({
				testID: 'cvv'
			})
			act(() => {
				cvvInput.props.onChangeText('123')
			})
			expect(cvvInput.props.value).toBe('123')
		})
		it('should not allow more than 3 digits', () => {
			const cvvInput = component.root.findByProps({
				testID: 'cvv'
			})
			act(() => {
				cvvInput.props.onChangeText('1234')
			})
			expect(cvvInput.props.value).toBe('123')
		})
	})
	describe('Pay Button', () => {
		it('should disable pay button if any input is empty', () => {
			const payButton = component.root.findByProps({
				testID: 'payButton'
			})
			expect(payButton.props.disabled).toBe(true)
		})

		it('should disable pay button if any input is invalid', () => {
			const payButton = component.root.findByProps({
				testID: 'payButton'
			})
			const cardNumberInput = component.root.findByProps({
				testID: 'cardNumber'
			})
			const cardHolderInput = component.root.findByProps({
				testID: 'cardHolder'
			})
			const expirationDateInput = component.root.findByProps({
				testID: 'expirationDate'
			})
			const cvvInput = component.root.findByProps({
				testID: 'cvv'
			})
			act(() => {
				cardNumberInput.props.onChangeText('1234567')
				cardHolderInput.props.onChangeText('John Doe')
				expirationDateInput.props.onChangeText('12')
				cvvInput.props.onChangeText('1')
			})
			expect(payButton.props.disabled).toBe(true)
		})

		it('should enable pay button if all inputs are filled correctly', () => {
			const payButton = component.root.findByProps({
				testID: 'payButton'
			})
			const cardNumberInput = component.root.findByProps({
				testID: 'cardNumber'
			})
			const cardHolderInput = component.root.findByProps({
				testID: 'cardHolder'
			})
			const expirationDateInput = component.root.findByProps({
				testID: 'expirationDate'
			})
			const cvvInput = component.root.findByProps({
				testID: 'cvv'
			})
			act(() => {
				cardNumberInput.props.onChangeText('1234567890123456')
				cardHolderInput.props.onChangeText('John Doe')
				expirationDateInput.props.onChangeText('1223')
				cvvInput.props.onChangeText('123')
			})
			expect(payButton.props.disabled).toBe(false)
		})
		// TODO: this is failing due to a bug in the test renderer
		// it('should show a modal with activity indicator on pay button press', () => {
		// 	act(() => {
		// 		component.unmount()
		// 		component = renderer.create(<ServicesAndProductsCheckout />)
		// 	})
		// 	expect(component.toJSON()).toMatchSnapshot()
		// 	const payButton = component.root.findByProps({
		// 		testID: 'payButton'
		// 	})
		// 	act(() => {
		// 		payButton.props.onPress()
		// 	})
		// 	const modal = component.root.findByProps({
		// 		testID: 'modalProcessing'
		// 	})
		// 	expect(modal.props.visible).toBe(true)
		// })
		// it('should hide modal on payment complete', async () => {
		// 	const payButton = component.root.findByProps({
		// 		testID: 'payButton'
		// 	})
		// 	await act(async () => {
		// 		payButton.props.onPress()
		// 		await Promise.resolve()
		// 	})
		// 	const modal = component.root.findByProps({
		// 		testID: 'modalProcessing'
		// 	})
		// 	expect(modal.props.visible).toBe(false)
		// })
		it('should call purchaseProduct on pay button press', async () => {
			const payButton = component.root.findByProps({
				testID: 'payButton'
			})
			await act(async () => {
				payButton.props.onPress()
				await Promise.resolve()
			})
			expect(useBusinessPartnerStore().purchaseProduct).toHaveBeenCalled()
		})
		it('should set the alert message on payment success', async () => {
			;(
				useBusinessPartnerStore()
					.purchaseProduct as unknown as jest.Mock
			).mockReturnValue({
				transaction_id: 'transaction_id',
				transaction_status: 'completed',
				transaction_date: 'transaction_date',
				message: 'message'
			})
			const payButton = component.root.findByProps({
				testID: 'payButton'
			})
			await act(async () => {
				payButton.props.onPress()
				await Promise.resolve()
			})
			expect(useAlertStore().setAlert).toHaveBeenCalledWith({
				type: 'success',
				message: 'productService.paymentSuccess'
			})
		})

		it('should set the alert message on payment failed', async () => {
			;(
				useBusinessPartnerStore()
					.purchaseProduct as unknown as jest.Mock
			).mockReturnValue({
				transaction_id: 'transaction_id',
				transaction_status: 'failed',
				transaction_date: 'transaction_date',
				message: 'message'
			})
			const payButton = component.root.findByProps({
				testID: 'payButton'
			})
			await act(async () => {
				component.update(<ServicesAndProductsCheckout />)
				payButton.props.onPress()
				await Promise.resolve()
			})
			expect(useAlertStore().setAlert).toHaveBeenCalledWith({
				type: 'error',
				message: 'productService.paymentFailed'
			})
		})

		it('should navigate back on payment complete', async () => {
			const payButton = component.root.findByProps({
				testID: 'payButton'
			})
			await act(async () => {
				payButton.props.onPress()
				await Promise.resolve()
			})
			expect(router.back).toHaveBeenCalled()
		})
	})
})
