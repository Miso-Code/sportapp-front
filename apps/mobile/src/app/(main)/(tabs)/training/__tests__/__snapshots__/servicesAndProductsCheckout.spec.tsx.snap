// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServicesAndProductsCheckout should render correctly 1`] = `
[
  <View
    testID="modalProcessing"
    visible={false}
  >
    <View
      style={
        {
          "alignItems": "center",
          "gap": 10,
          "justifyContent": "center",
        }
      }
    >
      <View
        accessibilityRole="progressbar"
        accessibilityState={
          {
            "busy": true,
          }
        }
        accessible={true}
        style={
          [
            {
              "alignItems": "center",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <View
          collapsable={false}
          style={
            {
              "height": 48,
              "opacity": 1,
              "width": 48,
            }
          }
        >
          <View
            collapsable={false}
            style={
              {
                "alignItems": "center",
                "bottom": 0,
                "justifyContent": "center",
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          >
            <View
              collapsable={false}
              style={
                {
                  "height": 48,
                  "transform": [
                    {
                      "rotate": "45deg",
                    },
                  ],
                  "width": 48,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  {
                    "height": 24,
                    "overflow": "hidden",
                    "width": 48,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    {
                      "height": 48,
                      "transform": [
                        {
                          "translateY": 0,
                        },
                        {
                          "rotate": "-165deg",
                        },
                      ],
                      "width": 48,
                    }
                  }
                >
                  <View
                    collapsable={false}
                    style={
                      {
                        "height": 24,
                        "overflow": "hidden",
                        "width": 48,
                      }
                    }
                  >
                    <View
                      collapsable={false}
                      style={
                        {
                          "borderColor": "rgba(103, 80, 164, 1)",
                          "borderRadius": 24,
                          "borderWidth": 4.8,
                          "height": 48,
                          "width": 48,
                        }
                      }
                    />
                  </View>
                </View>
              </View>
            </View>
          </View>
          <View
            collapsable={false}
            style={
              {
                "alignItems": "center",
                "bottom": 0,
                "justifyContent": "center",
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          >
            <View
              collapsable={false}
              style={
                {
                  "height": 48,
                  "transform": [
                    {
                      "rotate": "45deg",
                    },
                  ],
                  "width": 48,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  {
                    "height": 24,
                    "overflow": "hidden",
                    "top": 24,
                    "width": 48,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    {
                      "height": 48,
                      "transform": [
                        {
                          "translateY": -24,
                        },
                        {
                          "rotate": "345deg",
                        },
                      ],
                      "width": 48,
                    }
                  }
                >
                  <View
                    collapsable={false}
                    style={
                      {
                        "height": 24,
                        "overflow": "hidden",
                        "width": 48,
                      }
                    }
                  >
                    <View
                      collapsable={false}
                      style={
                        {
                          "borderColor": "rgba(103, 80, 164, 1)",
                          "borderRadius": 24,
                          "borderWidth": 4.8,
                          "height": 48,
                          "width": 48,
                        }
                      }
                    />
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <Text
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "writingDirection": "ltr",
            },
            [
              {
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": "400",
                "letterSpacing": 0.15,
                "lineHeight": 24,
              },
              undefined,
            ],
          ]
        }
      >
        Procesando el Pago
      </Text>
    </View>
  </View>,
  <RCTScrollView
    contentContainerStyle={
      {
        "alignItems": "center",
        "gap": 20,
        "justifyContent": "center",
        "marginHorizontal": 15,
        "paddingBottom": 110,
        "paddingTop": 110,
      }
    }
    contentInsetAdjustmentBehavior="automatic"
  >
    <View>
      <View
        category="category"
        image="image_url"
        price={100}
        priceFrequency="payment_frequency"
        testID="productCard"
        title="name"
      />
      <View>
        <View
          style={{}}
        >
          <View
            style={
              {
                "alignItems": "flex-start",
                "flexDirection": "row",
                "flexWrap": "wrap",
                "justifyContent": "flex-start",
                "marginBottom": 10,
                "marginTop": 10,
                "width": "100%",
              }
            }
          >
            <Text
              style={{}}
            >
              <Text
                style={
                  [
                    {},
                    {},
                  ]
                }
              >
                description
              </Text>
            </Text>
          </View>
        </View>
      </View>
      <TextInput
        keyboardType="numeric"
        label="Cantidad"
        mode="outlined"
        onChangeText={[Function]}
        style={
          {
            "width": "100%",
          }
        }
        testID="quantity"
        value="10"
      />
      <TextInput
        keyboardType="numeric"
        label="Card Number"
        left={
          <TextInput.Icon
            forceTextInputFocus={true}
            icon="credit-card-outline"
          />
        }
        mode="outlined"
        onChangeText={[Function]}
        style={
          {
            "width": "100%",
          }
        }
        testID="cardNumber"
      />
      <TextInput
        label="Card Holder"
        left={
          <TextInput.Icon
            forceTextInputFocus={true}
            icon="account"
          />
        }
        mode="outlined"
        onChangeText={[Function]}
        style={
          {
            "width": "100%",
          }
        }
        testID="cardHolder"
        value=""
      />
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "width": "100%",
          }
        }
      >
        <TextInput
          keyboardType="numeric"
          label="Expiration Date"
          left={
            <TextInput.Icon
              forceTextInputFocus={true}
              icon="calendar"
            />
          }
          mode="outlined"
          onChangeText={[Function]}
          testID="expirationDate"
        />
        <TextInput
          keyboardType="numeric"
          label="CVV/CVV2"
          left={
            <TextInput.Icon
              forceTextInputFocus={true}
              icon="lock"
            />
          }
          mode="outlined"
          onChangeText={[Function]}
          testID="cvv"
          value=""
        />
      </View>
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "rgba(28, 27, 31, 0.12)",
            "borderRadius": 5,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
            "width": "100%",
          }
        }
        testID="payButton-container-outer-layer"
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "rgba(28, 27, 31, 0.12)",
              "borderColor": "transparent",
              "borderRadius": 5,
              "borderStyle": "solid",
              "borderWidth": 0,
              "flex": undefined,
              "minWidth": 64,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="payButton-container"
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": true,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderRadius": 5,
                },
              ]
            }
            testID="payButton"
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  [
                    {
                      "textAlign": "left",
                    },
                    {
                      "color": "rgba(28, 27, 31, 1)",
                      "writingDirection": "ltr",
                    },
                    [
                      {
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      [
                        {
                          "marginHorizontal": 16,
                          "marginVertical": 9,
                          "textAlign": "center",
                        },
                        false,
                        {
                          "marginHorizontal": 24,
                          "marginVertical": 10,
                        },
                        undefined,
                        false,
                        {
                          "color": "rgba(28, 27, 31, 0.38)",
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        undefined,
                      ],
                    ],
                  ]
                }
                testID="payButton-text"
              >
                Pagar
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "transparent",
            "borderRadius": 5,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
            "width": "100%",
          }
        }
        testID="cancelButton-container-outer-layer"
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "transparent",
              "borderColor": "rgba(121, 116, 126, 1)",
              "borderRadius": 5,
              "borderStyle": "solid",
              "borderWidth": 1,
              "flex": undefined,
              "minWidth": 64,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="cancelButton-container"
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderRadius": 5,
                },
              ]
            }
            testID="cancelButton"
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  [
                    {
                      "textAlign": "left",
                    },
                    {
                      "color": "rgba(28, 27, 31, 1)",
                      "writingDirection": "ltr",
                    },
                    [
                      {
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      [
                        {
                          "marginHorizontal": 16,
                          "marginVertical": 9,
                          "textAlign": "center",
                        },
                        false,
                        {
                          "marginHorizontal": 24,
                          "marginVertical": 10,
                        },
                        undefined,
                        false,
                        {
                          "color": "rgba(103, 80, 164, 1)",
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        undefined,
                      ],
                    ],
                  ]
                }
                testID="cancelButton-text"
              >
                Cancelar
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>,
]
`;
